lsbi/model.py

```diff
@@ -1,236 +1,391 @@
+        A = np.broadcast_to(A, self.shape + (self.n, self.n))
+        D = dediagonalise(evidence.cov, evidence.diagonal, self.d)
+        D = np.broadcast_to(D, self.shape + (self.d, self.d))
+        C = np.einsum("...ja,...al->...jl", self._M, self._Sigma)
+        C = np.broadcast_to(C, self.shape + (self.d, self.n))
+        B = np.moveaxis(C, -1, -2)
+        Sigma = np.block([[A, B], [C, D]])
+        return multivariate_normal(mu, Sigma, self.shape, self.n + self.d)
+
+    @property
+    def _M(self):
+        return dediagonalise(self.M, self.diagonal_M, self.d, self.n)
+
+    @property
+    def _C(self):
+        return dediagonalise(self.C, self.diagonal_C, self.d)
+
+    @property
+    def _Sigma(self):
+        return dediagonalise(self.Sigma, self.diagonal_Sigma, self.n)
+
+
+class MixtureModel(LinearModel):
+    """A linear mixture model.
+
+    D|theta, A ~ N( m + M theta, C )
+    theta|A    ~ N( mu, Sigma )
+    A          ~ categorical( exp(logA) )
+
+    Defined by:
+        Parameters:          theta (..., n,)
+        Data:                D     (..., d,)
+        Prior means:         mu    (..., k, n)
+        Prior covariances:   Sigma (..., k, n, n)
+        Data means:          m     (..., k, d)
+        Data covariances:    C     (..., k, d, d)
+        log mixture weights: logA  (..., k,)
+
+    Parameters
+    ----------
+    M : array_like, optional
+        if ndim>=2: model matrices
+        if ndim==1: model matrix with vector diagonal for all components
+        if scalar: scalar * rectangular identity matrix for all components
+        Defaults to k copies of rectangular identity matrices
+    m : array_like, optional
+        if ndim>=1: data means
+        if scalar: scalar * unit vector for all components
+        Defaults to 0 for all components
+    C : array_like, optional
+        if ndim>=2: data covariances
+        if ndim==1: data covariance with vector diagonal for all components
+        if scalar: scalar * identity matrix for all components
+        Defaults to k copies of identity matrices
+    mu : array_like, optional
+        if ndim>=1: prior means
+        if scalar: scalar * unit vector for all components
+        Defaults to 0 for all components
+        Prior mean, defaults to zero vector
+    Sigma : array_like, optional
+        if ndim>=2: prior covariances
+        if ndim==1: prior covariance with vector diagonal for all components
+        if scalar: scalar * identity matrix for all components
+        Defaults to k copies of identity matrices
+    logA : array_like, optional
+        if ndim>=1: log mixture weights
+        if scalar: scalar * unit vector
+        Defaults to uniform weights
+    n : int, optional
+        Number of parameters, defaults to automatically inferred value
+    d : int, optional
+        Number of data dimensions, defaults to automatically inferred value
+    """
+
+    def __init__(self, logA=1, *args):
```

Comment:
Two tweaks:

*args, **kwargs

and

logA=1.0

Reply:
Many thanks -- ideally there would be a test for this (which hopefully doesn't expand the test suite too much), but before I do that, do you have any opinions on the API in general.
- Should LinearModel have positional arguments, or just kwargs? (the latter is easier to maintain)
- Are we happy with the notation? M,m,mu,C are short, to the point, and relatively standard, Sigma is a bit clunky, but S may not be standard enough. [pep8](https://peps.python.org/pep-3131/#policy-specification) has opinions on Σ and μ, but we could always allow users to provide Sigma and mu as kwargs (it becomes considerably easier with copilot to put in the greek!)

Reply:
How cursed does 31924fe look? (certainly benefits from the shorter variable names, but it would be a nightmare for users).

Reply:
How cursed does 31924fe look? (certainly benefits from the shorter variable names, but it would be a nightmare for users).